internal/uvm/vpmem_mapped.go

```diff
@@ -0,0 +1,307 @@
+package uvm
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Microsoft/hcsshim/internal/guestrequest"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/memory"
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	PageSize                = 0x1000
+	lcowPackedVPMemLayerFmt = "/run/layers/p%d-%d-%d"
+	MaxMappedDeviceCount    = 128
+)
+
+type mappedDeviceInfo struct {
+	vPMemInfoDefault
+	deviceOffset uint64
+	deviceSize   uint64
+}
+
+type vPMemInfoMulti struct {
+	memory.PoolAllocator
+	maxSize              uint64
+	maxMappedDeviceCount uint32
+	mappings             map[string]*mappedDeviceInfo
+}
+
+func newVPMemMappedDevice(hostPath, uvmPath string, deviceOffset, deviceSize uint64) *mappedDeviceInfo {
+	return &mappedDeviceInfo{
+		vPMemInfoDefault: vPMemInfoDefault{
+			hostPath: hostPath,
+			uvmPath:  uvmPath,
+			refCount: 1,
+		},
+		deviceOffset: deviceOffset,
+		deviceSize:   deviceSize,
+	}
+}
+
+func newPackedVPMemDevice() *vPMemInfoMulti {
+	return &vPMemInfoMulti{
+		PoolAllocator:        memory.NewPoolMemoryAllocator(),
+		maxSize:              DefaultVPMemSizeBytes,
+		mappings:             make(map[string]*mappedDeviceInfo),
+		maxMappedDeviceCount: MaxMappedDeviceCount,
+	}
+}
+
+func pageAlign(t uint64) uint64 {
+	if t%PageSize == 0 {
+		return t
+	}
+	return (t/PageSize + 1) * PageSize
+}
+
+// newMappedVPMemModifyRequest creates an hcsschema.ModifySettingsRequest to modify VPMem devices/mappings
+// for the multi-mapping setup
+func newMappedVPMemModifyRequest(ctx context.Context, rType string, deviceNumber uint32, md *mappedDeviceInfo, uvm *UtilityVM) (*hcsschema.ModifySettingRequest, error) {
+	request := &hcsschema.ModifySettingRequest{
+		RequestType: rType,
+	}
+	pmem := uvm.vpmemDevicesPacked[deviceNumber]
+	switch rType {
+	case requesttype.Add:
+		if pmem == nil {
+			request.Settings = hcsschema.VirtualPMemDevice{
+				ReadOnly:    true,
+				HostPath:    md.hostPath,
+				ImageFormat: "Vhd1",
+			}
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, deviceNumber)
+		} else {
+			request.Settings = hcsschema.VirtualPMemMapping{
+				HostPath:    md.hostPath,
+				ImageFormat: "Vhd1",
+			}
+			request.ResourcePath = fmt.Sprintf(vPMemDeviceResourceFormat, deviceNumber, md.deviceOffset)
+		}
+	case requesttype.Remove:
+		if pmem == nil {
+			return nil, errors.Errorf("no device found at location %d", deviceNumber)
+		}
+		if len(pmem.mappings) == 1 {
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, deviceNumber)
+		} else {
+			request.ResourcePath = fmt.Sprintf(vPMemDeviceResourceFormat, deviceNumber, md.deviceOffset)
+		}
+	default:
+		return nil, errors.New("unsupported request type")
+	}
+
+	request.GuestRequest = guestrequest.GuestRequest{
```

Comment:
Nit: Doesn't look like this relies on any work above or any branching logic so I'd prefer if we just set this up on line 68 with the request itself